base/delete_soon_helper.h

```diff
@@ -0,0 +1,52 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BASE_DELETE_SOON_HELPER_H_
+#define BRAVE_BASE_DELETE_SOON_HELPER_H_
+
+namespace base {
+// Helper that forces object to invalidate things before object destroyed
+// Subclass must implement DeleteSoonImpl and call DeleteSoon before destructor
+// of the object. Example usage would be like a object is a WeakPtr which is
+// called and destructed on different threads, in this case we can
+// DeleteSoonHelper to make sure the WeakPtr is invalidate on certain thread.
+// See tor::TorFileWatcher as a reference.
+template <typename T>
+class DeleteSoonHelper {
+ public:
+  constexpr DeleteSoonHelper() = default;
+  explicit DeleteSoonHelper(std::nullptr_t) = delete;
+
+  DeleteSoonHelper(const DeleteSoonHelper&) = delete;
+  DeleteSoonHelper& operator=(const DeleteSoonHelper&) = delete;
+
+  DeleteSoonHelper(DeleteSoonHelper&&) noexcept = default;
+  DeleteSoonHelper& operator=(DeleteSoonHelper&&) noexcept = default;
+
+  ~DeleteSoonHelper() { DCHECK(delete_soon_called); }
```

Comment:
curious, are you able to observe this DCHECK firing? I guess we at least should make the flag atomic to be able to observe it's change on another sequence.

Also, `DeleteSoon` is basically not guaranteed to complete, so to be honest I don't see much value in doing it - the fact that there was no DCHECK doesn't prove anything

Reply:
What do you mean "`DeleteSoon` is not guaranteed to complete? The purpose of the DCHECK is to alert that the object was destroyed without calling `DeleteSoon`

Reply:
it is not guaranteed that `SequencedTaskRunner::DeleteSoon` will actually post&execute the task. So it is not guaranteed that a destructor will be called , thus the fact that `DCHECK` doesn't fire doesn't mean that `DeleteSoon` was called

Reply:
addressed in https://github.com/brave/brave-core/pull/7927/commits/536325b0aca2613cdb3ca3e1c5deac2a25f7dfd9